src/webgpu/web_platform/reftests/canvas_composite_alpha.html.ts

```diff
@@ -0,0 +1,129 @@
+import { assert, unreachable } from '../../../common/util/util.js';
+
+import { runRefTest } from './gpu_ref_test.js';
+
+// <canvas> element from html page
+declare const cvs: HTMLCanvasElement;
+
+export function run(format: GPUTextureFormat, compositingAlphaMode: GPUCanvasCompositingAlphaMode) {
+  runRefTest(async t => {
+    const ctx = cvs.getContext('gpupresent');
+    assert(ctx !== null, 'Failed to get WebGPU context from canvas');
+
+    switch (format) {
+      case 'bgra8unorm':
+      case 'bgra8unorm-srgb':
+      case 'rgba8unorm':
+      case 'rgba8unorm-srgb':
+        break;
+      default:
+        unreachable();
+    }
+
+    const a = compositingAlphaMode === 'opaque' ? (1.0).toFixed(1) : (0.5).toFixed(1);
+
+    ctx.configure({
+      device: t.device,
+      format,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT,
+      compositingAlphaMode,
+    });
+
+    const pipeline = t.device.createRenderPipeline({
+      vertex: {
+        module: t.device.createShaderModule({
+          code: `
+struct VertexOutput {
+[[builtin(position)]] Position : vec4<f32>;
+[[location(0)]] fragColor : vec4<f32>;
+};
+
+[[stage(vertex)]]
+fn main([[builtin(vertex_index)]] VertexIndex : u32) -> VertexOutput {
+var pos = array<vec2<f32>, 6>(
+    vec2<f32>( 0.75,  0.75),
+    vec2<f32>( 0.75, -0.75),
+    vec2<f32>(-0.75, -0.75),
+    vec2<f32>( 0.75,  0.75),
+    vec2<f32>(-0.75, -0.75),
+    vec2<f32>(-0.75,  0.75));
+
+var offset = array<vec2<f32>, 4>(
+vec2<f32>( -0.125,  0.125),
+vec2<f32>( 0.125, 0.125),
+vec2<f32>(-0.125, -0.125),
+vec2<f32>( 0.125,  -0.125));
+
+var color = array<vec4<f32>, 4>(
+    vec4<f32>(0.49804, 0.0, 0.0, ${a}),
+    vec4<f32>(0.0, 0.49804, 0.0, ${a}),
+    vec4<f32>(0.0, 0.0, 0.49804, ${a}),
+    vec4<f32>(0.49804, 0.49804, 0.0, ${a})); // 0.49804 -> 0x7f
+
+var output : VertexOutput;
+output.Position = vec4<f32>(pos[VertexIndex % 6u] + offset[VertexIndex / 6u], 0.0, 1.0);
+output.fragColor = color[VertexIndex / 6u];
+return output;
+}
+        `,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        module: t.device.createShaderModule({
+          code: `
+[[stage(fragment)]]
+fn main([[location(0)]] fragColor: vec4<f32>) -> [[location(0)]] vec4<f32> {
+return fragColor;
+}
+        `,
+        }),
+        entryPoint: 'main',
+        targets: [
+          {
+            format,
+            blend:
+              compositingAlphaMode === 'opaque'
+                ? undefined
+                : {
+                    color: {
+                      srcFactor: 'src-alpha',
+                      dstFactor: 'one-minus-src-alpha',
+                      operation: 'add',
+                    },
+                    alpha: {
+                      srcFactor: 'one',
+                      dstFactor: 'one-minus-src-alpha',
+                      operation: 'add',
+                    },
+                  },
```

Comment:
Add comment about what this blend state is. IIUC it's converting unpremultiplied shader values to premultiplied texture values. Also can be shorter by taking advantage of defaults (maybe that's too confusing?)
```suggestion
                : {
                    color: { srcFactor: 'src-alpha' }, // Premultiply alpha into the color channels
                    alpha: {} // Pass through alpha
                  },
```

Reply:
Same here as above. We are drawing 4 quads in 4 different draw calls so the content blending inside the canvas context still matters (to match the result of canvas 2d ref tests)

Reply:
![image](https://user-images.githubusercontent.com/5031596/129816047-a66e1ece-4d30-47ac-9581-a98c2161f37f.png)
Canvas is 4x4
Drawing 3x3 quads with different color for each corner in order. (I should update with some comments as well)